### PR TITLE
[RFC][DNM] 2D Accelerator API Proposal

### PR DESCRIPTION
**Pull request by danieldegrasse**
## Introduction

This RFC proposes a generic API for interaction with 2D graphics accelerators.
This API is designed to integrate well with a graphics stack like LVGL, but it
can also be used directly by customer applications.

### Problem description

As customer applications continue to demand high performance graphics with
low power consumption, SOC vendors have begun to integrate 2D graphics IP
into their MCU lines. Zephyr currently has no layer to abstract graphics
operations, and leaves performance and power savings on the table without
a framework to enable this IP.

For example, NXP iMX RT MCUs contain a Pixel Processing Pipeline (2D DMA engine)
this engine must be leveraged to obtain acceptable performance when a customer
application rotates their display image. A subset of the functionality available
in this module can be enabled via the DMA API (as seen in #60064), but features
like dynamic rotation or blending multiple buffers are not available.

### Proposed change

The 2D accelerator API will provide a common block of operations, which may not
be supported by all graphics IP devices. Graphics stacks like LVGL may
implement support for acceleration using the 2D accelerator API, or customers
may program directly to the 2D accelerator API.

The following key operations should be supported by this API:
- Basic shape drawing
- Basic buffer blending and rotation
- Path based shape drawing
- Path based blending and rotation
- Pixel format conversion

The API is not intended to supersede the display API, but may replace the
`display_write` call in certain applications.

The usage model for an application would look similar to the following:
2d-broad-view

Note that the Graphics stack or user application is expected to fall back to
software rendering (which the 2D accelerator API does not implement) when
the accelerator device does not support a given operation.

### Command Queuing

GPU devices often enable queueing of multiple commands before running a render
to minimize CPU overhead, therefore all API functions are designed to enqueue
commands, which can then be flushed to start a render.

See an example below:
2d-standard-queue

Each render request will set up a command in the hardware's render queue.
`accel2d_render` will then run all queued operations, which will blend `Buffer0`
onto `Buffer1`. `Buffer1` can then be written to the display with
`display_write`. All render request functions are nonblocking.

### Capabilities API

In order to enable a graphics stack to decide which GPU operations it will
attempt to leverage the API for and which it will render using software,
an API will be provided to check GPU capabilities. This will reduce runtime
overhead, as graphics stacks will not call GPU APIs with unsupported
options only to discover software must be used for rendering after some
checks have been run by the driver.

### Basic APIs

For simple 2D DMA engines like the NXP PXP or ST DMA2D, only a limited
set of drawing commands can be supported. Exposing API signatures that these
IP blocks can implement will insure that they can be fully leveraged, and reduce
the runtime checks these drivers would need to ensure they could support a
given API call.

Therefore, a set of basic APIs will be available for GPU IP, including commands
for the following:
- Draw rectangle
- Draw arc
- Draw line
- Blend multiple images

### Path APIs

For more advanced vector graphics engines, APIs will be defined for drawing
arbitrary shapes. These API signatures will allow these IP blocks to be
fully leveraged within graphics stacks.

Path commands are similar to the format SVG uses to describe shapes, like
following:
paths

The following commands will be supported for this use case:
- Draw shape
- Copy image within bounds of shape

### Direct Display Output

Some graphics engines support outputting directly to a connected display,
without the need to stream the final render result back into a framebuffer
in memory. This can be combined with rotation, scaling, or pixel format
conversion. An API will be provided for this purpose, and when used it will
bypass `display_write`

### Multiple Accelerators

For some systems, a specific IP block may be better suited to raster/blit based
graphics tasks, while another IP block may be used for vector operations.
Another use case for this would be using one IP block for display
rotation, while another is enabled for rendering operations, like so:

```
gpu-shared: gpu-node {
	blit-engine = <&gc355>;
	vector-engine = <&gc355>;
	output-engine = <&pxp>; /* Used with accel2d_write_display */
};
```

The call graph for a GPU "Virtual GPU" implemented using this method might
look like the following when using LVGL:
2d-general-usage

## Detailed API Definition

For a detailed definition of the API proposal, please see the documentation
and header file present with this RFC. The documentation outlines the rationale
behind each 2D accelerator API, as well as which rendering engines would
support it.

## Dependencies

Any implementation of a 2D accelerator API will also need to be integrated with
LVGL in order to provide value for customers using Zephyr, as well as validate
the API is flexible enough to be useful. Zephyr's LVGL module revision
has support for accelerator backends with the merge of #59656. Therefore,
the majority of the work required would be integrating the 2D accelerator API
as an LVGL backend

### Alternatives

The primary alternative to supporting GPUs using a standardized API in Zephyr
would be to implement GPUs directly within LVGL. LVGL already has a
defined 2D acceleration backend: https://docs.lvgl.io/8.3/porting/gpu.html.

This backend lacks the ability to perform display rotation automatically,
but it is capable of leveraging the blending and vector graphics features
of the GPUs listed above. Support for automatic display rotation using
hardware is on the roadmap for the next major revision:
https://github.com/lvgl/lvgl/blob/master/docs/ROADMAP.rst#architecture

The primary advantage of this approach is that Zephyr would avoid
reinventing the wheel. There has already been significant work done to
optimize these accelerators within the LVGL backend, and using a GPU API within
Zephyr may only introduce an abstraction layer that reduces performance.

The primary disadvantages of using LVGL as the chosen method are that
Zephyr will be locked into LVGL for GPU support. Additionally, most drivers
will require some level of integration with devicetree and Zephyr OS primitives,
and without a driver class there will be more challenges as to where to
integrate these changes.

### Concerns and Unresolved Questions

The sampling of IP this proposal is based on is rather limited.
Currently, the IP I am aware of and working to support with this RFC includes
the following:

- NXP PXP (2D DMA engine)
  - scaling
  - rotation by 90/180/270 degrees
  - direct output to LCDIF peripheral
  - blending of multiple buffers
  - color space conversions

- ST DMA2D (2D DMA engine)
  - blending of multiple buffers
  - color space conversions

- NXP SmartDMA
  - rotation by 90/180/270 degrees
  - direct output to MIPI-DSI peripheral

- NXP GC355/GCNanoLiteV
  - Path based drawing (Gradient or solid fill)
  - blending of multiple buffers
  - arbitrary scaling, rotation, and translation of buffers

I have also looked at the ST Neochrom VG and Ambiq (Thinksilicon) NEMA Pico
GPU lines. Both appear to use path based drawing, and offer a similar feature
set to the GC355/GCNanoLiteV. However, given lack of available documentation I
cannot confirm the specifics of either of these GPUs.

### CODE REVIEW

**Comment by teburd** [comment]
What a great RFC, love all the diagrams and the well thought out ideas here.

**Comment by danieldegrasse** [comment]
> What a great RFC, love all the diagrams and the well thought out ideas here. In looking at the API and such I see some nice ideas around instantiating different commands (accel2d_op) type structures it looks like but I didn't see a way to submit a batch of operations to the accelerator in some manner, if that's a blocking/non-blocking operation, and how the results might be obtained.

Thanks for the feedback. One thing to consider is if there would be an efficiency gain from some form of API allowing you to queue operations like these in one call. I'm not currently aware of any 2D GPU device that can accept an array of commands like this, but it is something to consider (for example, the GPU might be able to perform a block of operations out of order when submitted in this manner)

File: include/zephyr/drivers/accel2d.h [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+	uint32_t y;
+};
+
+/**
+ * @brief 2D accelerator buffer descriptor
+ *
+ * Describes buffer for 2D accelerator. Note that some GPUs may have alignment
+ * restrictions on the stride or location of this buffer.
+ */
+struct accel_buf {
```
Comment (pdgendt) [inline include/zephyr/drivers/accel2d.h:59]: Naming similar to `display_` wrapped descriptor?

```suggestion
struct accel2d_buffer_descriptor {
```

File: include/zephyr/drivers/accel2d.h [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+	/** Buffer descriptor */
+	struct display_buffer_descriptor desc;
+};
+
+/**
+ * @brief 2D accelerator gradient direction
+ *
+ * 2D accelerator gradient direction. Gradient can be horizontal or vertical
+ */
+enum accel2d_grad_dir {
```
Comment (pdgendt) [inline include/zephyr/drivers/accel2d.h:73]: Any reason why `grad` is shortened here?

File: include/zephyr/drivers/accel2d.h [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+ * @brief 2D accelerator matrix definition
+ *
+ * 2D accelerator affine transformation matrix definition. This 3x2 matrix
+ * consists of a 2x2 rotation matrix, and a column vector for translation of the
+ * image. It can be used with the 2D accelerator path functions to perform
+ * arbitrary 2D transformations the 2D accelerator input buffer or output path.
+ */
+struct accel2d_matrix {
+	/** Transformation matrix definition */
+	float m[3][2];
```
Comment (pdgendt) [inline include/zephyr/drivers/accel2d.h:372]: Maybe a more descriptive member name, such as `transform` or `trans`?

File: include/zephyr/drivers/accel2d.h [hunk 4]
```diff
@@ -0,0 +1,680 @@
...
+
+/**
+ * @brief 2D accelerator capabilities flags
+ *
+ * Capabilities data for a 2D accelerator device.
+ */
+struct accel2d_capabilities {
+	/* Blending support */
+	unsigned int blend_support; /**< Bitmask of supported blending modes */
+	unsigned short alignment; /**< Alignment requirement for 2D accelerator buffers */
```
Comment (pdgendt) [inline include/zephyr/drivers/accel2d.h:435]: Any reason to not use fixed size types?

```suggestion
	uint32_t blend_support; /**< Bitmask of supported blending modes */
	uint16_t alignment; /**< Alignment requirement for 2D accelerator buffers */
```

File: include/zephyr/drivers/accel2d.h [hunk 5]
```diff
@@ -0,0 +1,680 @@
...
+ * @retval -ENOTSUP if the 2D accelerator does not support the
+ *          requested combination of settings
+ * @retval -EINVAL invalid buffer alignment settings for the 2D accelerator
+ * @retval -EBUSY @ref accel2d_render must be called before this operation can
+ *          proceed, as no more rendering can be queued on the 2D accelerator
+ * hardware.
+ * @retval -EIO Underlying 2D accelerator encountered an I/O error.
+ */
+int accel2d_draw_path(const struct device *dev, struct accel_buf *buf,
+		      struct gpu_src *src, uint32_t *path, uint32_t path_len,
```
Comment (pdgendt) [inline include/zephyr/drivers/accel2d.h:547]: The `struct gpu_src` isn't available? Maybe some leftover?

```suggestion
		      struct accel2d_src *src, uint32_t *path, uint32_t path_len,
```

File: include/zephyr/drivers/accel2d.h [hunk 6]
```diff
@@ -0,0 +1,680 @@
...
+ * @param rect: rectangle output description
+ * @param blend: 2D accelerator porter-duff blend setting
+ * @retval 0 command was queued successfully
+ * @retval -ENOSYS if this 2D accelerator does not support this feature
+ * @retval -ENOTSUP if the 2D accelerator does not support the
+ *          requested combination of settings
+ * @retval -EINVAL invalid buffer alignment settings for the 2D accelerator
+ * @retval -EBUSY @ref accel2d_render must be called before this operation can
+ *          proceed, as no more rendering can be queued on the 2D accelerator
+ * hardware.
```
Comment (pdgendt) [inline include/zephyr/drivers/accel2d.h:462]: Nit: Alignment, same for other API functions.

```suggestion
 * @retval -EBUSY @ref accel2d_render must be called before this operation can
 *          proceed, as no more rendering can be queued on the 2D accelerator
 *          hardware.
```

File: include/zephyr/drivers/accel2d.h [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+	/* Path API support */
+	unsigned int path_support: 1;   /**< support for accel2d path APIs */
+	unsigned int direct_support: 1; /**< support for accel2d_write_display */
+};
+
+/**
+ * @brief Draw a simple rectangle using the 2D accelerator
+ *
+ * Draws a simple rectangle using the 2D accelerator hardware.
+ * @param dev: 2D accelerator device object
```
Comment (pdgendt) [inline include/zephyr/drivers/accel2d.h:450]: Parameter documentation typically does not have a semicolon in zephyr. (same for other parameters/functions)

```suggestion
 * @param dev 2D accelerator device object
```

File: include/zephyr/drivers/accel2d.h [hunk 8]
```diff
@@ -0,0 +1,680 @@
...
+		      struct accel2d_capabilities *caps);
+
+/**
+ * @}
+ */
+
+#ifdef __cplusplus
+}
+#endif
+
```
Comment (pdgendt) [inline include/zephyr/drivers/accel2d.h:679]: Do we need to consider `CONFIG_USERSPACE`? The `display` API also doesn't use the `__syscall` attributes, just wondering if that is deliberate?

**Review by pdgendt — COMMENTED** [review]
Nice proposal!

**Comment by teburd** [comment]
> > What a great RFC, love all the diagrams and the well thought out ideas here. In looking at the API and such I see some nice ideas around instantiating different commands (accel2d_op) type structures it looks like but I didn't see a way to submit a batch of operations to the accelerator in some manner, if that's a blocking/non-blocking operation, and how the results might be obtained.
> 
> Thanks for the feedback. One thing to consider is if there would be an efficiency gain from some form of API allowing you to queue operations like these in one call. I'm not currently aware of any 2D GPU device that can accept an array of commands like this, but it is something to consider (for example, the GPU might be able to perform a block of operations out of order when submitted in this manner)

If you wish for the API to be used from userspace at all, with syscalls and everything, batching would bring quite a bit to the table. Imagine every vector drawing operation being a syscall and suddenly batching looks very appealing I imagine, even if the hardware doesn't have a command queue fifo software can provide it and feed it.

**Comment by fabiobaltieri** [comment]
Thanks @danieldegrasse, that's a really cool proposal. My main thought is about rewriting the current LVGL accelerator code in the Zephyr code base. Are you aware of other graphics API that could reuse this beyond  LVGL or other reason why the project is better off not using the LVGL code? From what I see right now there's no obvious winner, so in the end it may boil down to vendor support preferences, are the LVGL drivers maintained by the project authors or by NXP and ST folks?

**Comment by danieldegrasse** [comment]
> Are you aware of other graphics API that could reuse this beyond LVGL or other reason why the project is better off not using the LVGL code? From what I see right now there's no obvious winner, so in the end it may boil down to vendor support preferences, are the LVGL drivers maintained by the project authors or by NXP and ST folks?

I am not aware of any open source graphics APIs that could reuse this, but there are several graphical frameworks for MCUs that could potentially benefit from one abstraction layer in Zephyr. For example, emWin can leverage the PXP, ChromART (ST DMA2D), and Renesas D/AVE 2D core (see section 3.6 of the  [emWin user manual](https://www.segger.com/downloads/emwin/UM03001)).

Additionally, QT UltraLite supports the VGLite, Dave2D, DaveHD, PXP, and ChromART: https://doc.qt.io/QtForMCUs-2.5/qtul-overview.html

Although I don't know how these frameworks implement support for accelerators under the hood, I think there may be value in abstracting them at the Zephyr layer versus within LVGL.


> are the LVGL drivers maintained by the project authors or by NXP and ST folks?

For NXP, I believe we maintain our driver within LVGL. @nicusorcitu, what are your thoughts on this proposal?

**Comment by danieldegrasse** [comment]
Hi @nicusorcitu,

First off, thanks for engaging on this RFC- I appreciate the feedback.

> In LVGL v9, the task is basically the abstraction of the draw widgets which can be handled by CPU or by any other unit. Those tasks have a descriptor for every operation that can not be fully covered by your APIs.
E.G.
LVGL does not draw a simple line between two points. The line have a descriptor with width, dashed, rounded edges, opacity, etc
The rectangle can have rounded corners, have borders or outlines…

These are both good points- I expected to use the `accel2d_draw_path` API for most of the implementations in LVGL, and only use APIs like `accel2d_draw_rect` if the GPU did not support more advanced APIs, and if the shape LVGL requested did not require any additional features like rounded corners or borders.

> LVGL v9 will have support for vector API: https://github.com/lvgl/lvgl/issues/4388 It will have to be handled as individual task by VGLite.

Thanks for making me aware of this- this means that for a GPU like the VGLite, this RFC would effectively create an additional layer of abstraction on top of VGLite that was then just used as a backend for the LVGL vector API.

> I do believe the LVGL acceleration directly (without a Zephyr API layer) would be the best solution.
Adding an extra layer on top of the LVGL accelerators does not make much sense, further more I do not think it can support all the cases above and it will affect the overall performance.

Understood- then I think we need to consider what is needed in order to support a GPU in Zephyr without some form of 2D accelerator API like this. One option would be abstractions for the following:
- submitting operations to the GPU (IE queuing commands on VGLite)
- starting the GPU render
- waiting for render completion

The operations themselves could be something like a *very* simplified version of GEM command execution in DRM within Linux: https://docs.kernel.org/gpu/drm-mm.html?highlight=drm#command-execution. This way graphics operations can be device specific, and relatively flexible. What do you think about this approach?

**Comment by faxe1008** [comment]
Hi @danieldegrasse

Thanks a lot for the inital proposal, I watched your talk on the zd2023 :^).

> Understood- then I think we need to consider what is needed in order to support a GPU in Zephyr without some form of 2D accelerator API like this. One option would be abstractions for the following:
> 
> * submitting operations to the GPU (IE queuing commands on VGLite)
> * starting the GPU render
> * waiting for render completion

 Given then changes in LVGL v9 I think this is a reasonable way to continue. From what I gather from the current main `HEAD` is that each "drawing backend" can decide if it is capable to fullfil a certain draw task:
[lv_draw_sw.c](https://github.com/lvgl/lvgl/blob/master/src/draw/sw/lv_draw_sw.c#L77C17-L77C17)

Going with GEM here after reading through the spec seems like a good fit to solve this. One thing to consider is that GEM seems to also take care of client syncronization: [drm-mm.html#command-execution](https://docs.kernel.org/gpu/drm-mm.html?highlight=drm#command-execution) which from my understanding of the lvgl code is not needed since lvgl now wants to have an OSAL api to perform those operations. I think the Zephyr version of GEM does not need synchronization, instead the 3 points that you made would be sufficient.

Again the question when designing an API for this is how much we lean into the requirements of LVGL vs keeping the "driver features" generalistic. I think that covering most of the common use cases (blitting, filling, text rendering etc.) will already yield quite a big performance improvement and might be a good place to start.

PS.: I also looked at the OSAL requirements of LVGL and since they are quite simple, I will create an upstream PR to support Zephyr. The dependencies are quite limited and only target stable core APIs (`k_mutex`, `k_thread`, `k_cond`) so this should be reasonable to upstream. This will allow us to test any proposal for this current issue much more easily once v9 is released.

**Review by faxe1008 — CHANGES_REQUESTED** [review]
Proposal needs to be re-adjusted since the current primary use case (LVGL) changed its approach and the current RFC would introduce an additional abstraction layer that is not needed. As @danieldegrasse suggested going with an implementation that is inspired by established projects would be a good place to start.

Again, thanks for everyone participating in the discussion so far. I really enjoy seeing the zephyr graphics stack get some attention :^)

**Comment by nicusorcitu** [comment]
> Understood- then I think we need to consider what is needed in order to support a GPU in Zephyr without some form of 2D accelerator API like this. One option would be abstractions for the following:
> 
>     submitting operations to the GPU (IE queuing commands on VGLite)
>     starting the GPU render
>     waiting for render completion
> 

I do not really see the usability here:
- to submit any operation you need to call vg_lite_draw() or vg_lite_blit() APIs
- to start GPU render - just call vg_lite_flush()
- to wait for GPU completion - just call vg_lite_finish()

But it is up to you if you want even more abstraction.

**Comment by danieldegrasse** [comment]
> But it is up to you if you want even more abstraction.

I don't think we need any more abstraction at that level-the area I am focused on is how we integrate drivers like VGLite within Zephyr itself. Zephyr expects for drivers to use features like devicetree in order to determine the base address of a given peripheral, and I would prefer if we defined an API layer that enabled the lower end of the GPU driver (IE managing interrupts and writing to device registers) to be contained within Zephyr itself. Something along the lines of the way that the DRM framework in Linux works, although in this case the "UAPIs" for each GPU would be consumed by the graphics API (whether that be VGLite or something else) running on top of the accelerator.

**Comment by pdgendt** [comment]
Sad to see this got closed, is there any follow-up work planned?

**Comment by uLipe** [comment]
I'd like to reopen this RFC by giving an extra bunch of context from LVGL.

The draw units on the LVGL
currently do only platform specific code, but if you look on how these units are implemented they still need to call manufacturers low level commands and there is no standard for that.

For example the DMA2D assumes Cube Hal from ST, the PXP assumes MCUX , and VGLite depends on a kernel driver side that also has a port layer.

So if creating too abstract API for GPU can add a complexity layer, why not make simpler to connect LvGL draw units to GPUs using a minimal GPU command interface / Interrupt management?

Even the initial proposal is good since it would allow to create a single Zephyr draw unit on LVGL that could have multiple instances to use different acceleration methods if the SoC has them, or by having a single draw unit that can be on top of different MCU accelerators.

Also not only for LVGL case, having the GPU API would also make them general purpose usable for using for example on the imaging processing or even general computing.

**Comment by uLipe** [comment]
I can help on the continuation on it if there is interest on reopening.

**Comment by avolmat-st** [comment]
@uLipe I got to know this PR now. I'm very happy to see that you propose to help moving this forward, and I am fully aligned with your last comment (https://github.com/zephyrproject-rtos/zephyr/pull/60840#issuecomment-2899548674). This is exactly what I had in mind, for both LVGL usage but also using of those engines from within Zephyr for other imaging or other related stuff.

We have the following two PRs https://github.com/zephyrproject-rtos/zephyr/pull/111852 and https://github.com/zephyrproject-rtos/zephyr/pull/105970 in which we have Zephyr code in order to have those HW blocks working with LVGL. As part of it and since we had to describe those HW block within the dts, we introduced the bindings category ''graphic_accelerators''.
But as I wrote here (https://github.com/zephyrproject-rtos/zephyr/pull/111852#discussion_r3466288111) this is only to be considered as a first step in order to have those HW block usable on top of Zephyr / LVGL. Real target is to have a common 2D API so that after that we can have a unique LVGL (or any other graphic library) draw unit which interface to the Zephyr API in order to perform that.
Other Zephyr driver (say JPEG decoder / encoder) would also benefit from usage of DMA2D on STM32 so within Zephyr as well we need to be able to interface with those HW block.

Anyway, again, I think we are sharing same thought so I'd be happy to collaborate and help move this subject forward.

**Pull request closed without merging**